File: frontend/src/presentation/utils/superMultipla.ts
import type { LongshotCombo } from "@/presentation/utils/longshotCombos";

/** Promo Superbet — espelha config backend (SUPER_MULTIPLA_*). */
export const SUPER_MULTIPLA_MIN_LEG_ODD = 1.35;
export const SUPER_MULTIPLA_BONUS_PCT = 0.05;

export interface SuperMultiplaEnrichedCombo extends LongshotCombo {
  bonusEligible: boolean;
  bonusPercentage: number;
  finalReturn: number;
}

export interface SuperMultiplaCalculateLeg {
  id?: string;
  market: string;
  outcome: string;
  market_odd: number;
  model_prob?: number;
  superbet_event_id?: number;
  event_name?: string;
  selection_label?: string;
  is_live?: boolean;
  minute?: number;
}

export interface SuperMultiplaCalculateResult {
  slipId: string | null;
  totalOdds: number;
  productOdds: number | null;
  pricingMode: "product" | "bet_builder_sgm" | "bet_builder_heuristic";
  potentialPayout: number;
  bonusEligible: boolean;
  bonusPercentage: number;
  finalPayout: number;
  combinedProb: number | null;
  combinedEv: number | null;
  currency: string;
  warnings: string[];
  builderValidation: Record<string, unknown> | null;
  betType: "SIMPLE" | "MULTIPLE";
  legsCount: number;
}

function roundMoney(value: number): number {
  return Math.round(value * 100) / 100;
}

/** Elegibilidade promo quando todas as pernas têm odd >= 1.35. */
export function isSuperMultiplaBonusEligible(legs: { marketOdd: number }[]): boolean {
  return legs.length >= 2 && legs.every((leg) => leg.marketOdd >= SUPER_MULTIPLA_MIN_LEG_ODD);
}

/** Enriquece combo longshot com retorno final (+5% se elegível). */
export function applySuperMultiplaBonus(combo: LongshotCombo): SuperMultiplaEnrichedCombo {
  const bonusEligible = isSuperMultiplaBonusEligible(combo.legs);
  const bonusPercentage = bonusEligible ? SUPER_MULTIPLA_BONUS_PCT : 0;
  const finalReturn = bonusEligible
    ? roundMoney(combo.potentialReturn * (1 + SUPER_MULTIPLA_BONUS_PCT))
    : combo.potentialReturn;
  return {
    ...combo,
    bonusEligible,
    bonusPercentage,
    finalReturn,
  };
}

export function mapSuperMultiplaCalculate(raw: Record<string, unknown>): SuperMultiplaCalculateResult {
  return {
    slipId: (raw.slip_id as string | null) ?? null,
    totalOdds: Number(raw.total_odds),
    productOdds: raw.product_odds != null ? Number(raw.product_odds) : null,
    pricingMode:
      (raw.pricing_mode as SuperMultiplaCalculateResult["pricingMode"]) ?? "product",
    potentialPayout: Number(raw.potential_payout),
    bonusEligible: Boolean(raw.bonus_eligible),
    bonusPercentage: Number(raw.bonus_percentage ?? 0),
    finalPayout: Number(raw.final_payout),
    combinedProb: raw.combined_prob != null ? Number(raw.combined_prob) : null,
    combinedEv: raw.combined_ev != null ? Number(raw.combined_ev) : null,
    currency: String(raw.currency ?? "BRL"),
    warnings: Array.isArray(raw.warnings) ? (raw.warnings as string[]) : [],
    builderValidation: (raw.builder_validation as Record<string, unknown>) ?? null,
    betType: (raw.bet_type as "SIMPLE" | "MULTIPLE") ?? "MULTIPLE",
    legsCount: Number(raw.legs_count ?? 0),
  };
}

File: schemas/super_multipla.py
"""Schemas de Super Múltipla — cálculo de odds combinadas e bônus promo."""

from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator

from schemas.user_bet import PickInput


class SuperMultiplaLegInput(PickInput):
    """Perna de múltipla com contexto Superbet e ao vivo."""

    id: str | None = Field(None, description="UUID gerado pelo cliente")
    superbet_event_id: int | None = Field(None, description="ID do evento na Superbet")
    event_name: str | None = Field(None, description="Ex.: Brasil vs Marrocos")
    selection_label: str | None = Field(None, description="Rótulo exibido na Superbet")
    is_live: bool = Field(False, description="Perna capturada em jogo ao vivo")
    minute: int | None = Field(None, ge=0, le=120, description="Minuto no momento da seleção")

    @model_validator(mode="after")
    def _require_market_odd(self) -> "SuperMultiplaLegInput":
        if self.market_odd is None or self.market_odd < 1.01:
            raise ValueError("market_odd obrigatória e >= 1.01 por perna")
        return self


class SuperMultiplaEventContext(BaseModel):
    """Placar e minuto por evento (múltiplas cross-game)."""

    superbet_event_id: int
    minute: int | None = Field(None, ge=0, le=120)
    home_score: int = Field(0, ge=0)
    away_score: int = Field(0, ge=0)
    ht_home_score: int | None = Field(None, ge=0)
    ht_away_score: int | None = Field(None, ge=0)


class SuperMultiplaCalculateRequest(BaseModel):
    """Payload para cálculo stateless de slip."""

    legs: list[SuperMultiplaLegInput] = Field(..., min_length=1)
    stake: float = Field(..., gt=0, description="Valor apostado em R$")
    bet_type: Literal["SIMPLE", "MULTIPLE"] = "MULTIPLE"
    slip_id: str | None = None
    minute: int | None = Field(None, ge=0, le=120, description="Minuto global (jogo único)")
    home_score: int = Field(0, ge=0)
    away_score: int = Field(0, ge=0)
    ht_home_score: int | None = Field(None, ge=0)
    ht_away_score: int | None = Field(None, ge=0)
    superbet_event_id: int | None = Field(None, description="Evento principal quando slip single-game")
    event_contexts: list[SuperMultiplaEventContext] = Field(default_factory=list)


class SuperMultiplaCalculateResponse(BaseModel):
    """Resultado do cálculo de múltipla."""

    slip_id: str | None = None
    total_odds: float = Field(..., gt=1)
    product_odds: float | None = Field(None, gt=1, description="Produto simples das pernas (pré-SGM)")
    pricing_mode: str = Field(
        "product",
        description="product | bet_builder_sgm | bet_builder_heuristic",
    )
    potential_payout: float = Field(..., gt=0)
    bonus_eligible: bool = False
    bonus_percentage: float = Field(0.0, ge=0)
    final_payout: float = Field(..., gt=0)
    combined_prob: float | None = Field(None, ge=0, le=1)
    combined_ev: float | None = None
    currency: str = "BRL"
    warnings: list[str] = Field(default_factory=list)
    builder_validation: dict[str, Any] | None = None
    bet_type: Literal["SIMPLE", "MULTIPLE"] = "MULTIPLE"
    legs_count: int = Field(..., ge=1)
